csharp/ql/src/experimental/CWE-918/RequestForgery.qll

```diff
@@ -0,0 +1,202 @@
+    SystemWebHttpRequestMessageSink() {
+      exists(Class c | c.hasQualifiedName("System.Net.Http.HttpRequestMessage") |
+        c.getAConstructor().getACall().getArgument(1) = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * An argument to a `WebRequest.Create` call taken as a
+   * sink for Server Side Request Forgery(SSRF) Vulnerabilities. *
+   */
+  private class SystemNetWebRequestCreateSink extends Sink {
+    SystemNetWebRequestCreateSink() {
+      exists(Method m |
+        m.getDeclaringType().hasQualifiedName("System.Net.WebRequest") and m.hasName("Create")
+      |
+        m.getACall().getArgument(0) = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * An argument to a new HTTP Request call of a `System.Net.Http.HttpClient` object
+   * taken as a sink for Server Side Request Forgery(SSRF) Vulnerabilities.
+   */
+  private class SystemNetHttpClientSink extends Sink {
+    SystemNetHttpClientSink() {
+      exists(Method m |
+        m.getDeclaringType().hasQualifiedName("System.Net.Http.HttpClient") and
+        m.hasName([
+            "DeleteAsync", "GetAsync", "GetByteArrayAsync", "GetStreamAsync", "GetStringAsync",
+            "PatchAsync", "PostAsync", "PutAsync"
+          ])
+      |
+        m.getACall().getArgument(0) = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * An url argument to a method call of a `System.Net.WebClient` object
+   * taken as a sink for Server Side Request Forgery(SSRF) Vulnerabilities.
+   */
+  private class SystemNetClientBaseAddressSink extends Sink {
+    SystemNetClientBaseAddressSink() {
+      exists(Property p |
+        p.hasName("BaseAddress") and
+        p.getDeclaringType()
+            .hasQualifiedName(["System.Net.WebClient", "System.Net.Http.HttpClient"])
+      |
+        p.getAnAssignedValue() = this.asExpr()
+      )
+    }
+  }
+
+  /**
+   * A method call which checks the base of the tainted uri is assumed
+   * to be a guard for Server Side Request Forgery(SSRF) Vulnerabilities.
+   * This guard considers all checks as valid.
+   */
+  private class BaseUriGuard extends SanitizerGuard, MethodCall {
+    BaseUriGuard() {
+      this.getTarget().getDeclaringType().hasQualifiedName("System.Uri") and
+      this.getTarget().hasName("IsBaseOf")
+    }
+
+    override predicate checks(Expr e, AbstractValue v) {
+      e = this.getArgument(0) and
+      v.(AbstractValues::BooleanValue).getValue() = true
+    }
+  }
+
+  private class StringFormatSanitizer extends Sanitizer {
+    StringFormatSanitizer() {
+      exists(FormatCall c, int i | c.getArgument(i) = this.asExpr() | i > 0)
+    }
+  }
+
+  private class PathCombineSanitizer extends Sanitizer {
+    PathCombineSanitizer() {
+      exists(MethodCall combineCall, int i |
+        combineCall.getTarget().hasQualifiedName("System.IO.Path", "Combine") and
+        combineCall.getArgument(i) = this.asExpr()
+      |
+        i > 0
+      )
+    }
+  }
+
+  private class InterpolatedStringSanitizer extends Sanitizer {
```

Comment:
This check is not working the way I expect it to work. @tamasvajk Can you please take a look?
I am trying to block all flows from the interpolated string except when 
1. It is of the form ``$"http://{`taint`}/blabla/");"``
2. It is something like `$"{`taint`}/blabla/");"`
3. Or when it is used in an unrelated call such as `System.Console.WriteLine($"Getting {taint}");`


However, my current version fails to accommodate for particular two cases which result in 2 false negatives in the result set.

The first occurs in `microsoft/Atlas`. The query fails to alert here as the url parameter is marked sanitized when it is not. I tried changing `InterpolatedStringSanitizer` from a `Sanitizer` to `SanitizerIn` or `SanitizerOut` but niether of them worked. I don't know whats wrong here.

The other two occur in `ddogadkiney/csharp-netcore-sca-02` and `nopSolutions/nopCommerce`. The code base for both of them is quite similar. Seems like both of them are derived from the same code base, The false negative here is due to 
```
$"{(useSsl ? Uri.UriSchemeHttps : Uri.UriSchemeHttp)}{Uri.SchemeDelimiter}{hostHeader.FirstOrDefault()}";
```
Due to the way I have written my Sanitizer for interpolated strings, I can't deal with these type of cases. Considering the fact that the comparison in this case could have been done separately not as part of the interpolated statement, Hard coding the pattern seems bad. Hence, I am left with two alternatives here. Leave these results out of the query or else add a second data flow to consider flow to and from an interpolated string. Considering the fact that adding a flow comes with a heavy performance penalty, I am in for either hard coding the pattern or just using local flow(in this case results would be lost obviously but it's much more performance efficient) or leave the results as is. WDYT?

Reply:
I wouldn't care about the second case, where the uri scheme is not hard coded, but flows somehow into the string. It would make the query unnecessarily complex to handle this fully as the delimiter could also flow from a variable or not...

Can you give a link to the false negative in `microsoft/Atlas`?

Reply:
@tamasvajk https://github.com/microsoft/Atlas/blob/8056fead2b122a711f0e611bd7c0e6b43148c87b/src/Microsoft.Atlas.CommandLine/Execution/WorkflowEngine.cs#L194-L215 
The `_console.WriteLine` on L204 sanitizes the request.url somehow.If you comment out the sanitizer, the alert comes back again.

Reply:
Sorry, I still can't repro this. I commented out the sanitizer, but the issue is still not showing up: https://lgtm.com/query/6556373478803825645/

What am I missing?

Reply:
@tamasvajk The alert is shown on the link you shared. Are you sure you can't see an alert there?

Reply:
I see one alert, but it's not the one you linked above in `WorkflowEngine.cs`.

Reply:
Sorry, I probably didn't check the path in detail to see that this is actually the issue that you linked above. Coming back to sanitizer: if I don't comment it out, I still get the same result: https://lgtm.com/query/3356943433915439332/ 
So there's no false negative here, is there?

Reply:
That's interesting. I am trying to run it on my local node and I can still experience the issue. Is the LGTM running on the latest main?

Reply:
Okay so no I cleared my codeql cache and tried again. it works fine. I can see the alert here. I guess i will mark this as resolved then.